### DISCUSSION TITLE
Torchio datasets & histograms

### DISCUSSION THREAD

**Discussion by stavrizinonos** [TorchIO-project/torchio#572] [Q&A] [answered]
Does Torchio dataset include 'International Neuroimaging Data-Sharing Initiative (INDI)' data with brain MRI scans? (documentation: http://fcon_1000.projects.nitrc.org/). It contains 1200+ ‘resting state’ functional MRI (R-fMRI) datasets independently collected at 33 sites, and it would be very useful if it was available to be used directly. If not, how can it be processed using Torchio's functionalities? Also, how are the 'landmarks.npy' created based on the voxels from the mri and brain scans?

**Comment by fepegar** [comment, marked as answer]
Hi, @stavrizinonos. This dataset is not available in TorchIO at the moment. Does it need registration? You are free to work on this, if you would like to see it in TorchIO.

About the histogram standardization, I recommend you take a look at the documentation: https://torchio.readthedocs.io/transforms/preprocessing.html#histogramstandardization

**Reply by stavrizinonos** [reply]
Thank you @fepegar for your answer. I have downloaded the FCP dataset from their page. Its accessible publicly. My question now is if i can use a dataset not provided from Torchio but use its functionalities, for example apply transformations on it etc. Also, not about how to standardize the histograms but how to create them and pass them into a .npy file in the first place.

**Reply by fepegar** [reply]
Yes, you can do that. Please take a look at the "General" tutorial and let me know if you still have questions: https://github.com/fepegar/torchio/tree/master/notebooks#general

**Reply by stavrizinonos** [reply]
Thank you, amazing tutorial. Though, i would like to understand some things about the preprocessing of the dataset. Are there any steps that show how robex and the nifty registration to MNI are used in this google collab tutorial? Because,i have tried transforming my images to MNI space from Torchio documentations but is seems theres a problem finding the affine matrix. I have used the Collin dataset and tried to include it as reference space but still i cant get the transformation done. I really don’t know what im getting wrong as i am following the tutorials as they are.
Thank you in advance

**Reply by fepegar** [reply]
> Thank you, amazing tutorial.

Thanks!

> Though, i would like to understand some things about the preprocessing of the dataset. Are there any steps that show how robex and the nifty registration to MNI are used in this google collab tutorial?

I can share the script I used if you like.

> Because,i have tried transforming my images to MNI space from Torchio documentations but is seems theres a problem finding the affine matrix. I have used the Collin dataset and tried to include it as reference space but still i cant get the transformation done. I really don’t know what im getting wrong as i am following the tutorials as they are.

Can you share some code? What did you do? What did you expect? What did you get? You can follow the _Report a bug_ template in the Issues tab of this repo.

**Reply by stavrizinonos** [reply]
Hey, i would love to if you could share the script as it would be very helpful!
For the template registration i tried to follow this example from this documentation https://torchio.readthedocs.io/_/downloads/en/latest/pdf/ , page 37. but the problem with this is that it cannot find the transform_to_mni.txt in order to read the matrix, and also from this one https://colab.research.google.com/github/fepegar/torchio-notebooks/blob/main/notebooks/Data_preprocessing_and_augmentation_using_TorchIO_a_tutorial.ipynb#scrollTo=9F7Vikrer2Fn, the following part:

mni = tio.datasets.Colin27()
to_mni = tio.Resample(mni.t1.path, pre_affine_name='affine_matrix')
transformed=to_mni(tio.ScalarImage('scan_mprage_anonymized.nii.gz'))

And I receive this error "ValueError: An affine name was given ("affine_matrix"), but it was not found in any image in the subject"

Also in terms of ROBEX i following the steps from their GitHub account, installed pyrobex and then went with this in my notebook:

import nibabel as nib
from pyrobex.robex import robex
image = nib.load('scan_mprage_anonymized.nii.gz')
stripped, mask = robex(image)

But,  i keep getting 'OSError: [WinError 193] %1 is not a valid Win32 application'

**Reply by fepegar** [reply]
> Hey, i would love to if you could share the script as it would be very helpful!

This is the code I used once all the images and their corresponding brain segmentations had been transformed into the MNI space.

```python
import sys
from tqdm import tqdm
from pathlib import Path

import numpy as np
import SimpleITK as sitk

from utils import sglob, AffineMatrix


def get_sigma(scaling):
    """
    Compute optimal standard deviation for Gaussian kernel.

        Cardoso et al., "Scale factor point spread function matching:
        beyond aliasing in image resampling", MICCAI 2015
    """
    k = np.array(scaling)
    variance = (k ** 2 - 1 ** 2) * (2 * np.sqrt(2 * np.log(2))) ** (-2)
    sigma = np.sqrt(variance)
    return sigma


dataset_name, output_dir = sys.argv[1:3]

# Orient Slicer ILP, spacing 2.18,4.13,3.95
reference_path = '/tmp/colin_tiny.nii.gz'
reference = sitk.ReadImage(str(reference_path))
scaling = 2.18, 4.13, 3.95
sigmas = get_sigma(scaling)

dataset_dir = Path('~/datasets').expanduser() / dataset_name
mni_dir = dataset_dir / 'mni'
brain_dir = dataset_dir / 'brain'
output_dir = Path(output_dir)
output_image_dir = output_dir / 'image'
output_label_dir = output_dir / 'label'
output_image_dir.mkdir(exist_ok=True, parents=True)
output_label_dir.mkdir(exist_ok=True, parents=True)

identity = sitk.Transform()

mnis = sglob(mni_dir, '*.nii.gz')
for mni_path in tqdm(mnis):
    stem = mni_path.stem.split('_t1_pre')[0]
    l = sglob(brain_dir, f'{stem}*')
    trsf_path = sglob(mni_dir, f'{stem}*.txt')[0]
    if not l: continue
    brain_path = l[0]

    transform_to_mni = AffineMatrix(trsf_path).get_itk_transform()

    image = sitk.ReadImage(str(mni_path))
    brain = sitk.ReadImage(str(brain_path))

    image = sitk.DiscreteGaussian(image, sigmas)

    image = sitk.Resample(image, reference, identity, sitk.sitkLinear)
    brain = sitk.Resample(
        brain, reference, transform_to_mni, sitk.sitkNearestNeighbor)

    image_path = output_image_dir / f'{stem}_image.nii.gz'
    label_path = output_label_dir / f'{stem}_label.nii.gz'

    sitk.WriteImage(image, str(image_path))
    sitk.WriteImage(brain, str(label_path))
```

Please open a new discussion if you have further questions, as this one is about "datasets & histograms". It will the project organized!

**Reply by fepegar** [reply]
This is `utils`: https://github.com/fepegar/utils
